rust/kernel/net/ethtool.rs

```diff
@@ -0,0 +1,231 @@
+            &NetDevice::<T>::from_pointer(dev),
+            &mut EthToolTsInfo::from_pointer(info)
+        )?;
+        Ok(0)
+    }
+}
+
+pub(crate) struct EthToolOperationsVtable<T: NetDeviceAdapter>(marker::PhantomData<T>);
+
+impl<T: NetDeviceAdapter> EthToolOperationsVtable<T> {
+    const VTABLE: bindings::ethtool_ops = bindings::ethtool_ops {
+        _bitfield_align_1: [],
+        _bitfield_1: bindings::__BindgenBitfieldUnit::<[u8; 1usize]>::new([0u8; 1usize]),
+        supported_coalesce_params: 0,
+        get_drvinfo: if T::EthOps::TO_USE.get_drvinfo {
+            Some(get_drvinfo_callback::<T>)
+        } else {
+            None
+        },
+        get_regs_len: None,
+        get_regs: None,
+        get_wol: None,
+        set_wol: None,
+        get_msglevel: None,
+        set_msglevel: None,
+        nway_reset: None,
+        get_link: None,
+        get_link_ext_state: None,
+        get_eeprom_len: None,
+        get_eeprom: None,
+        set_eeprom: None,
+        get_coalesce: None,
+        set_coalesce: None,
+        get_ringparam: None,
+        set_ringparam: None,
+        get_pause_stats: None,
+        get_pauseparam: None,
+        set_pauseparam: None,
+        self_test: None,
+        get_strings: None,
+        set_phys_id: None,
+        get_ethtool_stats: None,
+        begin: None,
+        complete: None,
+        get_priv_flags: None,
+        set_priv_flags: None,
+        get_sset_count: None,
+        get_rxnfc: None,
+        set_rxnfc: None,
+        flash_device: None,
+        reset: None,
+        get_rxfh_key_size: None,
+        get_rxfh_indir_size: None,
+        get_rxfh: None,
+        set_rxfh: None,
+        get_rxfh_context: None,
+        set_rxfh_context: None,
+        get_channels: None,
+        set_channels: None,
+        get_dump_flag: None,
+        get_dump_data: None,
+        set_dump: None,
+        get_ts_info: if T::EthOps::TO_USE.get_ts_info {
+            Some(get_ts_info_callback::<T>)
+        } else {
+            None
+        },
+        get_module_info: None,
+        get_module_eeprom: None,
+        get_eee: None,
+        set_eee: None,
+        get_tunable: None,
+        set_tunable: None,
+        get_per_queue_coalesce: None,
+        set_per_queue_coalesce: None,
+        get_link_ksettings: None,
+        set_link_ksettings: None,
+        get_fecparam: None,
+        set_fecparam: None,
+        get_ethtool_phy_stats: None,
+        get_phy_tunable: None,
+        set_phy_tunable: None,
+    };
+
+    /// Builds an instance of [`struct ethtool_ops`].
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the adapter is compatible with the way the device is registered.
+    pub(crate) const unsafe fn build() -> &'static bindings::ethtool_ops {
+        &Self::VTABLE
+    }
+}
+
+/// Represents which fields of [`struct ethtool_ops`] should pe populated with pointers for the trait [`EthToolOps`]
+pub struct EthToolToUse {
```

Comment:
This also seems to be pretty common for bindings. Maybe we need to have a attribute macro. I'll think about a sensible design.